sdks/java/io/neo4j/src/main/java/org/apache/beam/sdk/io/neo4j/Neo4jIO.java

```diff
@@ -0,0 +1,1221 @@
+ * PCollection} to the graph by collecting a batch of elements after which all elements in the batch
+ * are written together to Neo4j.
+ *
+ * <p>Like the source, to configure this sink, you have to provide a {@link DriverConfiguration}.
+ *
+ * <p>In the following example we'll merge a collection of {@link org.apache.beam.sdk.values.Row}
+ * into Person nodes. Since this is a Sink it has no output and as such no RowMapper is needed. The
+ * rows are being used as a container for the parameters of the Cypher statement. The used Cypher in
+ * question needs to be an UNWIND statement. Like in the read case, the parameters {@link
+ * SerializableFunction} converts parameter values to a {@link Map}. The difference here is that the
+ * resulting Map is stored in a {@link List} (containing maps) which in turn is stored in another
+ * Map under the name provided by the {@link WriteUnwind#withUnwindMapName(String)} method. All of
+ * this is handled automatically. You do need to provide the unwind map name so that you can
+ * reference that in the UNWIND statement.
+ *
+ * <p>
+ *
+ * <p>For example:
+ *
+ * <pre>{@code
+ * pipeline
+ *   .apply(...)
+ *   .apply(Neo4jIO.<Row>writeUnwind()
+ *      .withDriverConfiguration(Neo4jIO.DriverConfiguration.create("neo4j://localhost:7687", "neo4j", "password"))
+ *      .withUnwindMapName("rows")
+ *      .withCypher("UNWIND $rows AS row MERGE(n:Person { id : row.id } ) SET n.firstName = row.first, n.lastName = row.last")
+ *      .withParametersFunction( row -> ImmutableMap.of(
+ *        "id", row.getString("id),
+ *        "first", row.getString("firstName")
+ *        "last", row.getString("lastName")))
+ *    );
+ * }</pre>
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class Neo4jIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Neo4jIO.class);
+
+  /**
+   * Read all rows using a Neo4j Cypher query.
+   *
+   * @param <ParameterT> Type of the data representing query parameters.
+   * @param <OutputT> Type of the data to be read.
+   */
+  public static <ParameterT, OutputT> ReadAll<ParameterT, OutputT> readAll() {
+    return new AutoValue_Neo4jIO_ReadAll.Builder<ParameterT, OutputT>().build();
+  }
+
+  /**
+   * Write all rows using a Neo4j Cypher UNWIND cypher statement. This sets a default batch size of
+   * 5000.
+   *
+   * @param <ParameterT> Type of the data representing query parameters.
+   */
+  public static <ParameterT> WriteUnwind<ParameterT> writeUnwind() {
+    return new AutoValue_Neo4jIO_WriteUnwind.Builder<ParameterT>()
+        .setBatchSize(ValueProvider.StaticValueProvider.of(5000L))
+        .build();
+  }
+
+  private static <ParameterT, OutputT> PCollection<OutputT> getOutputPCollection(
+      PCollection<ParameterT> input,
+      DoFn<ParameterT, OutputT> writeFn,
+      @Nullable Coder<OutputT> coder) {
+    PCollection<OutputT> output = input.apply(ParDo.of(writeFn));
```

Comment:
Unless writeFn (should be readFn ?) is a Splittable DoFn this means that each input element (source definition) is not splittable at all hence will be read from a single machine. This could be a significant performance drawback for many users.